radix-transactions/src/manifest/static_resource_movements_visitor.rs

```diff
@@ -0,0 +1,2171 @@
+
+    pub fn output(
+        self,
+    ) -> (
+        IndexMap<ComponentAddress, Vec<AccountDeposit>>,
+        IndexMap<ComponentAddress, Vec<AccountWithdraw>>,
+    ) {
+        (self.account_deposits, self.account_withdraws)
+    }
+}
+
+impl ManifestInterpretationVisitor for StaticResourceMovementsVisitor {
+    type Error<'a> = StaticResourceMovementsError<'a>;
+
+    // region:Invocation
+    fn on_start_instruction<'a>(
+        &mut self,
+        OnStartInstruction { index, effect }: OnStartInstruction<'a>,
+    ) -> ControlFlow<Self::Error<'a>> {
+        // We only care about invocations. Ignore anything that is not an invocation.
+        let ManifestInstructionEffect::Invocation { kind, args } = effect else {
+            return ControlFlow::Continue(());
+        };
+
+        // Getting all of the buckets and expressions in the arguments.
+        let (buckets, expressions) = {
+            let encoded = match manifest_encode(args) {
+                Ok(encoded) => encoded,
+                Err(error) => {
+                    return ControlFlow::Break(
+                        ManifestValidationError::ArgsEncodeError(error).into(),
+                    )
+                }
+            };
+            let mut traverser = ManifestTraverser::new(
+                &encoded,
+                ExpectedStart::PayloadPrefix(MANIFEST_SBOR_V1_PAYLOAD_PREFIX),
+                VecTraverserConfig {
+                    max_depth: MANIFEST_SBOR_V1_MAX_DEPTH,
+                    check_exact_end: true,
+                },
+            );
+            let mut buckets = index_set_new();
+            let mut expressions = index_set_new();
+            loop {
+                let event = traverser.next_event();
+                match event.event {
+                    TraversalEvent::TerminalValue(value) => match value {
+                        TerminalValueRef::Custom(ManifestCustomTerminalValueRef(
+                            ManifestCustomValue::Bucket(bucket),
+                        )) => {
+                            buckets.insert(bucket);
+                        }
+                        TerminalValueRef::Custom(ManifestCustomTerminalValueRef(
+                            ManifestCustomValue::Expression(expression),
+                        )) => {
+                            expressions.insert(expression);
+                        }
+                        _ => {}
+                    },
+                    TraversalEvent::ContainerStart(_)
+                    | TraversalEvent::ContainerEnd(_)
+                    | TraversalEvent::TerminalValueBatch(_) => {}
+                    TraversalEvent::DecodeError(error) => {
+                        return ControlFlow::Break(
+                            ManifestValidationError::ArgsDecodeError(error).into(),
+                        );
+                    }
+                    TraversalEvent::End => break,
+                }
+            }
+            (buckets, expressions)
```

Comment:
(Demonstrated in https://github.com/radixdlt/radixdlt-scrypto/pull/1931, the core visitor code is down to ~300 lines with the change, have a look and see what you think)

Reply:
Yup, I know that it can be done with the `OnEndInstruction`. To me it seems like different ways of doing the same thing. I do not necessarily think that one is better than the other.